src/util/getNextThoughtsWithContextChain.js

```diff
@@ -0,0 +1,178 @@
+      const firstChild = firstChildOfThought(derivedCursor || RANKED_ROOT, showHiddenThoughts)
+      if (firstChild) {
+        return { nextThoughts: unroot(cursor.concat(firstChild)), contextChain: [] }
+      }
+    }
+
+    const derivedContextRanked = (contextChain.length === 1 && isValidContextView(contextChain[0])) ? contextChain[0] : contextRanked
+    const derivedContext = derivedContextRanked.map(c => c.value)
+    const nextSibling = nextSiblingContext(value, rank, derivedContext, thoughtIndex, showHiddenThoughts)
+    const contextRankedHead = head(derivedContextRanked)
+
+    if (nextSibling) {
+      return { nextThoughts: [...nextSibling, contextRankedHead], contextChain: contextOf(contextChain) }
+    }
+    // jump out
+    return thoughtViewLookUp(contextRankedHead.value, contextOf(derivedContext), contextRankedHead.rank, contextOf(cursor), contextOf(contextChain), true)
+  }
+  const contextWithNoChildren = contextChain.length === 1 && isContextViewActive(contextChain[0]) && getContextsSortedAndRanked(head(pathToContext(contextChain[0]))).length <= 1
+  return thoughtViewLookUp(value, context, rank, cursor, contextChain, contextWithNoChildren)
+}
+
+const thoughtViewLookUp = (value, context, rank, cursor, contextChain, ignoreChildren) => {
+  const { showHiddenThoughts } = store.getState()
+
+  if (contextChain.length > 1 && head(contextChain).length === 1) {
+    return
+  }
+  const derivedCursor = !ignoreChildren && contextChain.length > 1 ? getDerivedCursor(contextChain) : cursor
+  const derivedContextRanked = contextOf(derivedCursor)
+  // map is expensive than duplicate condition check hence using the former
+  const derivedContext = !ignoreChildren && contextChain.length > 1 ? contextOf(derivedCursor).map(c => c.value) : context
+
+  const firstChild = firstChildOfThought(derivedCursor || RANKED_ROOT, showHiddenThoughts)
+  if (!ignoreChildren && firstChild) {
+    return { nextThoughts: unroot(cursor.concat(firstChild)), contextChain: [] }
+  }
+  const sortPreference = getSortPreference(meta(pathToContext(derivedContext)))
+  const siblings = sortPreference === 'Alphabetical' ? getThoughtsSorted(derivedContext) : getThoughtsRanked(derivedContext)
+
+  const contextHead = ((contextChain).length > 1 && head(contextChain).length < 3) ? head(contextChain)[0].value : head(derivedContext)
+  if (siblings.length === 0) {
+    return getNextThoughtsWithContextChain(contextHead, contextOf(derivedContext), rank, contextOf(derivedCursor), derivedContextRanked, contextChain.length > 1 ? [contextChain.slice(-1)] : contextChain)
+  }
+
+  const nextSibling = thoughtNextSibling(value, derivedContext, rank, showHiddenThoughts)
+  if (nextSibling) {
+    return { nextThoughts: unroot(derivedContextRanked.concat(nextSibling)), contextChain: contextOf(contextChain) }
+  }
+  // look for uncle/ancestor sibling
+  else {
+    const parentThought = derivedCursor.slice(-2, -1)[0]
```

Comment:
Do not use array methods on paths or thoughts. That is breaking the separation between interface and implementation. Use the interface methods, `head` and `contextOf`.

This is a really important rule that you should apply everywhere. Directly operating on the array implementation is going to make it impossible to change the implementation in the future. We should reply on the API that is exposed through the available methods.

Reply:
I agree with you on that. And I've updated such instances. 
We'll need a util function to return the first item in the list of thoughts/contexts to ger rid of all array operations.